lnrpc/routerrpc/router_server.go

```diff
@@ -401,11 +456,299 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+		MaxParts:         1,
+		AllowSelfPayment: false,
+		AmtMsat:          amtMsat,
+		PaymentHash:      paymentHash[:],
+		FeeLimitSat:      routeFeeLimitSat,
+		PaymentAddr:      payReq.PaymentAddr[:],
+		FinalCltvDelta:   int32(payReq.MinFinalCLTVExpiry()),
+	}
+
+	hints := payReq.RouteHints
+	isLsp, lspAdjustedRouteHints, lspHopHint := isLSP(hints, amtMsat)
+	if !isLsp {
+		probeRequest.RouteHints = invoicesrpc.CreateRPCRouteHints(hints)
+		return s.sendProbePayment(ctx, probeRequest)
+	}
+
+	if len(lspAdjustedRouteHints) == 0 && lspHopHint == nil {
+		return nil, errors.New("invalid lsp route hints")
+	}
+
+	// The adjusted route hints serve the payment probe to find the last
+	// public hop to the LSP on the route.
+	probeRequest.Dest = lspHopHint.NodeID.SerializeCompressed()
+	probeRequest.RouteHints = invoicesrpc.CreateRPCRouteHints(
+		lspAdjustedRouteHints,
+	)
+
+	// Add the CLTV expiry delta of the LSP hop hint to the payment probe.
+	probeRequest.FinalCltvDelta += int32(lspHopHint.CLTVExpiryDelta)
+
+	// The payment probe will be able to calculate the fee up until the LSP
+	// node. The fee of the last hop has to be calculated manually. Since
+	// the last hop's fee amount has to be sent across the payment path we
+	// have to add it to the original payment amount. Only then will the
+	// payment probe be able to determine the correct fee to the last hop
+	// prior to the private destination. For example, if the user wants to
+	// send 1000 sats to a private destination and the last hop's fee is 10
+	// sats, then 1010 sats will have to arrive at the last hop. This means
+	// that the probe has to be dispatched with 1010 sats to correctly
+	// calculate the routing fee.
+	//
+	// Calculate the hop fee for the last hop manually.
+	hopFee, err := lspHopHint.HopFee(int64(*payReq.MilliSat))
+	if err != nil {
+		return nil, err
+	}
+
+	// Add the last hop's fee to the requested payment amount that we want
+	// to get an estimate for.
+	probeRequest.AmtMsat += hopFee
+
+	// Dispatch the payment probe with adjusted fee amount.
+	resp, err := s.sendProbePayment(ctx, probeRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	// The probe succeeded, so we can add the last hop's fee to the
+	// to fee the payment probe returned.
+	resp.RoutingFeeMsat += hopFee
+
+	return resp, nil
+}
+
+// isLSP checks if the route hints indicate an LSP. An LSP is indicated in the
+// first return value if the last node in each route hint has the same node id.
+// If an LSP is indicated, the remaining hops in each route hint are returned
+// to the caller to be considered by the caller as route hints to the LSP node.
+// Additionally, the identified LSP node is returned as a separate hop hint.
+// This hop hint also contains the maximum fee setting if LSP hop hints contain
+// different fee settings. This is to obtain an upper bound of the fee estimate.
+func isLSP(routeHints [][]zpay32.HopHint, amtMsat int64) (bool,
+	[][]zpay32.HopHint, *zpay32.HopHint) {
```

Comment:
nonblocking: Following this double array of HopHints is a bit tough, I might suggest using types to package one of the layers into something like "Route" and implement some functions on that type to do the stripping you do in the body of this function.

Reply:
Initially I was looking for a `Route` struct, but once familiar with it, I found the implicit design nice. Adding a new type here but not in other places would also confuse, unless we changed it in all places.

Reply:
done